src/primaite/agents/aegis/modules/llm.py

```diff
@@ -80,100 +83,231 @@ def apply_chat_template(self, system: str, prompt: str, close_usr_msg: bool = Fa
 
     def _concat_graph_tags(self, graph_embs: torch.Tensor) -> torch.Tensor:
         """Given graph embeddings, concatenate the start and end tags <graph> ... </graph>"""
-        return torch.cat([self.graph_start_emb, graph_embs, self.graph_end_emb], dim=1)
+        batch_size = graph_embs.shape[0]
+        return torch.cat(
+            [
+                self.graph_start_inputs["inputs_embeds"].expand(batch_size, -1, -1),
+                graph_embs,
+                self.graph_end_inputs["inputs_embeds"].expand(batch_size, -1, -1),
+            ],
+            dim=1,
+        )
+
+    def format_inputs(self, inputs: torch.Tensor, graph_embeddings: torch.Tensor) -> torch.Tensor:
+        assert inputs["inputs_embeds"].shape[0] == graph_embeddings.shape[0]
+        batch_size = inputs["inputs_embeds"].shape[0]
+        original_len = inputs["inputs_embeds"].shape[1]
+
+        embeds = inputs["inputs_embeds"]
+        attention_mask = inputs["attention_mask"]
+        # Concatenate all sequences up to the text prompt embeddings
+        embeds = torch.cat(
+            [
+                self.start_msg_inputs["inputs_embeds"].expand(batch_size, -1, -1),
+                graph_embeddings,
+                embeds,
+            ],
+            dim=1,
+        )
+
+        # How many new tokens did we add to the beginning of the sequence?
+        n_tokens = embeds.shape[1] - original_len
+        attention_mask = torch.cat([torch.ones((batch_size, n_tokens)), attention_mask], dim=1)
+
+        # Expand all sequences to have minimum of close_msg_inputs tokens as slack
```

Comment:
I did this before I knew that it doesn't matter about their being pad tokens in between bits of text - but no point undoing the work xD